mrmustard/physics/representations.py

```diff
@@ -0,0 +1,305 @@
+
+    def from_ansatz(self, ansatz: Ansatz) -> Representation:
+        r"""
+        Returns a representation object from an ansatz object.
+        To be implemented by subclasses.
+        """
+        raise NotImplementedError
+
+    def __eq__(self, other: Representation) -> bool:
+        r"""
+        Whether this representation is equal to another.
+        """
+        return self.ansatz == other.ansatz
+
+    def __add__(self, other: Representation) -> Representation:
+        r"""
+        Adds this representation to another.
+        """
+        return self.from_ansatz(self.ansatz + other.ansatz)
+
+    def __sub__(self, other) -> Representation:
+        r"""
+        Subtracts another representation from this one.
+        """
+        return self.from_ansatz(self.ansatz - other.ansatz)
+
+    def __mul__(self, other: Representation | Scalar) -> Representation:
+        r"""
+        Multiplies this representation by another or by a scalar.
+        """
+        try:
+            return self.from_ansatz(self.ansatz * other.ansatz)
+        except AttributeError:
+            return self.from_ansatz(self.ansatz * other)
+
+    def __rmul__(self, other: Representation | Scalar) -> Representation:
+        r"""
+        Multiplies this representation by another or by a scalar on the right.
+        """
+        return self.__mul__(other)
+
+    def __truediv__(self, other: Representation | Scalar) -> Representation:
+        r"""
+        Divides this representation by another or by a scalar.
+        """
+        try:
+            return self.from_ansatz(self.ansatz / other.ansatz)
+        except AttributeError:
+            return self.from_ansatz(self.ansatz / other)
+
+    def __rtruediv__(self, other: Representation | Scalar) -> Representation:
+        r"""
+        Divides this representation by another or by a scalar on the right.
+        """
+        return self.from_ansatz(other / self.ansatz)
+
+    def __and__(self, other: Representation) -> Representation:
+        r"""
+        Takes the outer product of this representation with another.
+        """
+        return self.from_ansatz(self.ansatz & other.ansatz)
+
+
+class Bargmann(Representation):
+    r"""This class is the Fock-Bargmann representation of a broad class of quantum states,
+    transformations, measurements, channels, etc.
+
+    The ansatz available in this representation is a linear combination of
+    exponentials of bilinear forms with a polynomial part:
+
+    .. math::
+        F(z) = \sum_i \textrm{poly}_i(z) \textrm{exp}(z^T A_i z / 2 + z^T b_i)
+
+    This function allows for vector space operations on Bargmann objects including
+    linear combinations, outer product, and inner product.
+    The inner product is defined as the contraction of two Bargmann objects across
+    marked indices. This can also be used to contract existing indices
+    in a single Bargmann object, e.g. to implement the partial trace.
+
+    Examples:
```

Comment:
I don't think this will render on the docs. Could you please replace this with `code-block` and remove the ``>>>``?

Reply:
It was meant for doctests, but I haven't had time to try (we should consider using it). I'll use a code block for now.

Reply:
Oh I see!
Looks similar to the one I was consider introducing, nbsphinx:
https://pypi.org/project/nbsphinx/

Well one way or the other, great that we're all working towards better docs :)